ultravox/tools/infer_tool.py

```diff
@@ -190,25 +191,47 @@ def dataset_infer(inference: base.VoiceInference, args: InferArgs):
     if args.seed is not None:
         ds_args.shuffle_seed = args.seed
     ds = datasets.create_dataset(args.data_sets[0], ds_args)
-    scores: List[float] = []
-    for i, sample in enumerate(datasets.Range(ds, args.num_samples)):
-        # Store the original question and answer for JSON output.
-        question_text = sample.audio_transcript
-        expected_answer = sample.messages[-1]["content"]
-        # Drop any assistant response from the sample.
-        sample.messages = sample.messages[:-1]
-        if not args.json:
+
+    if args.json and isinstance(inference, infer.LocalInference):
+        if not args.batch_size:
+            args.batch_size = 1
+
+        start_time = time.time()
+        current_batch = []
+        for i, sample in enumerate(datasets.Range(ds, args.num_samples)):
+            current_batch.append(sample)
```

Comment:
An issue with creating the batch manually (as opposed to using a PyTorch `DataLoader`) is that everything will happen on a single thread and that your GPUs will be waiting for the data to get ready.
The `DataLoader` on the other hand uses thread(s) in the background to make sure the data is always needed when the GPU needs it.

I'm okay with not making the change right now (since that might make keeping `expected_answer` harder), but let's at least put a TODO or comment here that explains that there is room for improvement.

Reply:
@farzadab is there a way to use DataLoader without using a data_collator (or using a dummy data_collator), or some other type of DataLoader that uses multithreads for loading data?

Reply:
You can just use a pass-through `collate_fn`, e.g. `torch.utils.data.DataLoader(ds, batch_size=4, collate_fn=lambda x: x)`:

```python
from ultravox.data import datasets
import torch.utils.data
ds = datasets.LibriSpeechDummyDataset(datasets.VoiceDatasetArgs())
dl = torch.utils.data.DataLoader(ds, batch_size=4, collate_fn=lambda x: x)
print(next(iter(dl)))
```

Outputs:
```
[VoiceSample(messages=[{'role': 'user', 'content': 'Transcribe\n<|audio|>'}, {'role': 'assistant', 'content': 'Mister Quilter is the apostle of the middle classes and we are glad to welcome his gospel'}], audio=array([...
...
]
```

Reply:
I've added in the dataloader now